src/trs/hmtrs.rs

```diff
@@ -34,144 +92,106 @@ pub struct HMTRS {
-                let head_type = self.infer_op(op)?.instantiate(ctx).apply(ctx);
-                if ctx.unify(&head_type, &body_type).is_ok() {
-                    Ok(self.infer_op(op)?.instantiate(ctx).apply(ctx))
-                } else {
-                    Err(TypeError)
-                }
+                let head_type = self.op_tp(op)?.instantiate(ctx).apply(ctx);
+                let body_type = {
+                    let mut pre_types = Vec::with_capacity(args.len() + 1);
+                    for a in args {
+                        pre_types.push(self.infer_internal(a, ctx)?);
+                    }
+                    pre_types.push(ctx.new_variable());
+                    Type::from(pre_types)
+                };
+                ctx.unify(&head_type, &body_type)?;
+                Ok(self.op_tp(op)?.instantiate(ctx).apply(ctx))
             }
         }
     }
-    // half the internal recursive workhorse of type inference.
-    fn infer_args(&self, args: &[Term], ctx: &mut Context) -> Result<Type, TypeError> {
-        let mut pre_types = vec![];
-        for a in args {
-            pre_types.push(self.infer_internal(a, ctx)?);
-        }
-        pre_types.push(ctx.new_variable());
-        Ok(Type::from(pre_types))
-    }
-    /// Infer the [`TypeSchema`] of a [`Rule`] and the corresponding [`Context`] or generate a [`TypeError`].
+    /// Infer the [`TypeSchema`] of a [`Rule`], along with schemas for the left-hand-side and
+    /// right-hand-sides of the rule.
     ///
-    /// [`Context`]: ../../polytype/struct.Context.html
     /// [`Rule`]: ../../term_rewriting/struct.Rule.html
     /// [`TypeSchema`]: ../../polytype/struct.TypeSchema.html
-    /// [`TypeError`]: struct.TypeError.html
     pub fn infer_rule(
         &self,
         r: &Rule,
         ctx: &mut Context,
     ) -> Result<(TypeSchema, TypeSchema, Vec<TypeSchema>), TypeError> {
         let lhs_schema = self.infer_term(&r.lhs, ctx)?;
         let lhs_type = lhs_schema.instantiate(ctx);
-        let mut rhs_types = vec![];
-        let mut rhs_schemas = vec![];
-        for rhs in r.rhs.iter() {
+        let mut rhs_types = Vec::with_capacity(r.rhs.len());
+        let mut rhs_schemas = Vec::with_capacity(r.rhs.len());
+        for rhs in &r.rhs {
             let rhs_schema = self.infer_term(&rhs, ctx)?;
             rhs_types.push(rhs_schema.instantiate(ctx));
             rhs_schemas.push(rhs_schema);
         }
         for rhs_type in rhs_types {
-            ctx.unify(&lhs_type, &rhs_type).or(Err(TypeError))?;
+            ctx.unify(&lhs_type, &rhs_type)?;
         }
         // Get the variables bound by the lexicon
         let bound_vs = self.free_vars()
             .iter()
-            .flat_map(|x| {
-                ctx.substitution()
-                    .get(x)
-                    .unwrap_or(&Type::Variable(*x))
-                    .vars()
-            })
+            .flat_map(|x| Type::Variable(*x).apply(&ctx).vars())
```

Comment:
We use this `|x| Type::Variable(*x).apply(&ctx).vars()` construction a couple times. Is it worth naming and refactoring into its own function?

Reply:
I don't think it's worth refactoring because the intent is very clear and it's quite short.